src/util/string.h

```diff
@@ -17,6 +18,54 @@
 #include <vector>
 
 namespace util {
+/** Type to denote a format string that was checked at compile time */
+template <unsigned num_params>
+struct ConstevalFormatString {
+    const char* const fmt;
+    consteval ConstevalFormatString(const char* str) : fmt{str} { Detail_CheckNumFormatSpecifiers(fmt); }
+    constexpr static void Detail_CheckNumFormatSpecifiers(std::string_view str)
+    {
+        unsigned count_nor{0}; // Number of "normal" specifiers, like %s
+        unsigned count_pos{0}; // Max number in positional specifier, like %8$s
+        bool begin_f{false};
+        for (auto it{str.begin()}; it < str.end(); ++it) {
+            bool perc{*it == '%'};
+            if (begin_f) {
+                // Almost any char is allowed by tinyformat to be used in a
+                // format specifier, but a stricter check avoids bugs. For
+                // instance, a strict check prevents consuming the newline
+                // terminator in the format string "%\n". Also, excluding '0'
+                // avoids an out-of-range position error in tinyformat for the
+                // format string "%0$s\n".
+                if (std::string_view{"123456789aAcdfeEfFgGinopsuxX%"}.find(*it) == std::string_view::npos) throw "Invalid format specifier!";
```

Comment:
We do have quite a few use cases of `\.[\d+]f` precision that would throw here. No issue in the current code, but would like have to be overhauled in the next PR?